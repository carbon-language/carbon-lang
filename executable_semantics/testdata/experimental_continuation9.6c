// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
// Exceptions. See /LICENSE for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

// Test the way in which copying of continuations interacts with data
// on the stack such as the variable `x`. The copy of the continuation
// happens after the variable `x` is created, so both continuations
// refer to the same `x`. The first continuation finishes and deallocates
// `x`, then the second continuation finishes and tries to deallocate
// `x` again, which is a runtime error.

fn main() -> Int {
  __continuation k1 {
    var Int: x = 0;
    __await;
  }
  __run k1;
  var __Continuation: k2 = k1;
  __run k1;
  __run k2;
  return 0;
}
