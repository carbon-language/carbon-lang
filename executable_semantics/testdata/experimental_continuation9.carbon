// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
// Exceptions. See /LICENSE for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

package ExecutableSemanticsTest api;

// Test that the read from x triggers an error because x is dead.
// This test also demonstrates how by-reference free-variable capture
// is dangerous and can happen inside continuations.

fn capture() -> __Continuation {
  var x: i32 = 1;
  __continuation k {
    var y: i32 = x;
  }
  return k;
}

fn main() -> i32 {
  var k: __Continuation = capture();
  __run k; // error, lifetime of x is over
  return 0;
}
