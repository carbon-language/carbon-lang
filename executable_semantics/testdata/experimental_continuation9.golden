RUNTIME ERROR: executable_semantics/testdata/experimental_continuation9.carbon:14: undefined behavior: access to dead value 1
EXIT CODE: 255
